python/xir/program.py

```diff
@@ -211,165 +213,245 @@ class XIRProgram:
-            set[str]: all functions as strings contained in a set
+            Iterator[str]: free parameter variables as strings
         """
-        return self._called_ops
+        return iter(self._variables)
 
-    def add_gate(self, name: str, params: List[str], wires: Tuple, statements: List[Statement]):
-        """Adds a gate to the program
+    def add_called_function(self, name: str) -> None:
+        """Adds the name of a called function to the XIR program.
+
+        Args:
+            name (str): name of the function
+        """
+        self._called_functions.add(name)
+
+    def add_declaration(self, key: str, decl: Declaration) -> None:
+        """Adds a declaration to the XIR program.
+
+        Args:
+            key (str): key of the declaration ('func', 'gate', 'operator', or 'output')
+            decl (Declaration): the declaration
+        """
+        if key not in self._declarations:
+            raise KeyError(f"Key '{key}' is not a supported declaration.")
+
+        if decl.name in (decl.name for decl in self._declarations[key]):
+            warnings.warn(f"{key.title()} '{decl.name}' has already been declared.")
+
+        self._declarations[key].append(decl)
+
+    def add_gate(
+        self,
+        name: str,
+        params: Sequence[str],
+        wires: Sequence[Wire],
+        statements: Sequence[Statement],
+    ) -> None:
+        """Adds a gate to the XIR program.
 
         Args:
             name (str): name of the gate
-            params (str): parameters used in the gate
-            wires (str): wires that the gate is applied to
-            statements (list[Statement]): statements that the gate applies
+            params (Sequence[str]): parameters used in the gate
+            wires (Sequence[Wire]): wires that the gate is applied to
+            statements (Sequence[Statement]): statements that the gate applies
         """
         if name in self._gates:
-            warnings.warn("Gate already defined. Replacing old definition with new definiton.")
+            warnings.warn(
+                f"Gate '{name}' already defined. Replacing old definition with new definition."
+            )
+
         self._gates[name] = {"params": params, "wires": wires, "statements": statements}
 
+    def add_include(self, include: str) -> None:
+        """Adds an included XIR module to the XIR program.
+
+        Args:
+            include (str): name of the XIR module
+        """
+        if include in self._includes:
+            warnings.warn(f"Module '{include}' is already included. Skipping include.")
+            return None
+
+        self._includes.append(include)
+
     def add_operator(
-        self, name: str, params: List[str], wires: Tuple, statements: List[OperatorStmt]
-    ):
-        """Adds an operator to the program
+        self,
+        name: str,
+        params: Sequence[str],
+        wires: Sequence[Wire],
+        statements: Sequence[OperatorStmt],
+    ) -> None:
```

Comment:
I would probably just leave out the `-> None` here (same as not having a `Return None` by the end of this method). Would your opinion be it's preferable to have this?

Reply:
Fair point! 💯